addr_manager.go

```diff
@@ -161,60 +220,218 @@ func (mgr *AddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Durat
 	mgr.addrmu.Lock()
 	defer mgr.addrmu.Unlock()
-	mgr.init()
 
-	mgr.addrs[p] = make(addrSet) // clear what was there before
+	mgr.cache.Add(p, make(addrSet))
+	mgr.deleteAddrs(p)
+}
+
+func (mgr *AddrManager) warmupCache(p peer.ID) (addrSet, error) {
+	cached, found := mgr.cache.Get(p)
+	if found {
+		return cached.(addrSet), nil
+	}
+	stored, err := mgr.loadAddrs(p)
+	if err != nil {
+		return nil, err
+	}
+	mgr.cache.Add(p, stored)
+	return stored, nil
+}
+
+func mkExpAddrKey(p peer.ID, addr ma.Multiaddr) ds.Key {
+	return ds.NewKey(peerNamespace).Instance(peer.IDB58Encode(p)).
+		ChildString(addrNamespace).Instance(base32.StdEncoding.EncodeToString(addr.Bytes())).
+		ChildString(expNamespace)
+}
+
+func parseExpAddrKey(key ds.Key) (peer.ID, ma.Multiaddr, error) {
+	if key.Name() != expNamespace {
+		return "", nil, addrParseErr
+	}
+	if key = key.Parent(); key.Type() != addrNamespace {
+		return "", nil, addrParseErr
+	}
+	addrBytes, err := base32.StdEncoding.DecodeString(key.Name())
+	if err != nil {
+		return "", nil, addrParseErr
+	}
+	addr, err := ma.NewMultiaddrBytes(addrBytes)
+	if err != nil {
+		return "", nil, addrParseErr
+	}
+	if key = key.Parent(); key.Type() != peerNamespace || !key.IsTopLevel() {
+		return "", nil, addrParseErr
+	}
+	p, err := peer.IDB58Decode(key.Name())
+	if err != nil {
+		return "", nil, addrParseErr
+	}
+	return p, addr, nil
+}
+
+func parseExpAddr(key ds.Key, value interface{}) (peer.ID, expiringAddr, error) {
+	p, addr, err := parseExpAddrKey(key)
+	if err != nil {
+		return "", expiringAddr{}, err
+	}
+	expBytes, ok := value.([]byte)
+	if !ok {
+		return "", expiringAddr{}, addrParseErr
+	}
+	var expUnixNano int64
+	reader := bytes.NewBuffer(expBytes)
+	if err = binary.Read(reader, binary.LittleEndian, &expUnixNano); err != nil {
+		return "", expiringAddr{}, addrParseErr
+	}
+	return p, expiringAddr{Addr: addr, TTL: time.Unix(0, expUnixNano)}, nil
+}
+
+func (mgr *AddrManager) storeAddr(p peer.ID, addr ma.Multiaddr, expire time.Time) error {
+	dsk := mkExpAddrKey(p, addr)
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.LittleEndian, expire.UnixNano()); err != nil {
+		return err
+	}
+	return mgr.dstore.Put(dsk, buf.Bytes())
+}
+
+func (mgr *AddrManager) loadAddrs(p peer.ID) (addrSet, error) {
+	dsk := ds.NewKey(peerNamespace).Instance(peer.IDB58Encode(p)).ChildString(addrNamespace)
+	res, err := mgr.dstore.Query(dsq.Query{Prefix: dsk.String()})
+	if err != nil {
+		return nil, err
+	}
+	addrEnts, err := res.Rest()
+	if err != nil {
+		return nil, err
+	}
+	addrs := make(addrSet, len(addrEnts))
+	for _, addrEnt := range addrEnts {
+		_, addr, err := parseExpAddr(ds.NewKey(addrEnt.Key), addrEnt.Value)
+		if err != nil {
+			log.Warningf("failed to parse expiring address from datastore (key = %s): %s", addrEnt.Key, err)
+			continue
+		}
+		addrs[addr.Addr.String()] = addr
+	}
+	return addrs, nil
+}
+
+func (mgr *AddrManager) deleteAddr(p peer.ID, addr ma.Multiaddr) error {
```

Comment:
Callers of deleteAddr actually clear cache entries. I renamed functions that touch only datastore to be more explicit about it.

I added the test just in case.